%\VignetteIndexEntry{The RTandem users guide}
%\VignetteKeywords{MassSpectrometry, Proteomics}
%\VignettePackage{RTandem}
\documentclass[11pt]{article}
\usepackage{hyperref}
\usepackage{url}
\usepackage{color, pdfcolmk}
\usepackage[authoryear,round]{natbib}
\bibliographystyle{plainnat}


\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}

\author{Frederic Fournier\footnote{frederic.fournier@crchuq.ulaval.ca} and Arnaud Droit\footnote{arnaud.droit@crchuq.ulaval.ca}}

\begin{document}
\title{RTandem: An R encapsulation of X!Tandem}
\maketitle



\textnormal {\normalfont}
Introduction to RTandem

\tableofcontents
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newpage


\section{Licensing}

This package and the underlying X!Tandem code are distributed under the Artistic license 1.0. You are free to use and redistribute this software. 

\section{Introduction}

X!Tandem is an open source software for protein identification from tandem mass spectrometry experiments, and RTandem is an R encapsulation of this software. As now, RTandem provides the most basic possible encapsulation of X!Tandem: it has a single function that takes as argument the path to a parameter file. This function calls X!Tandem on the parameter file. The results are written 'outside of R', in a file whose location is listed in the parameters while information on the computation progress is sent to the R console. 

\section{Manifesto}

Right now, most analytic pipelines for protein identification using tandem mass spectrometry experiments are proprietary. Those pipelines often have the advantage of being user-friendly (to a certain extent), but reliance on proprietary pipelines has many downfalls. Proprietary pipelines hinder collaboration between labs by making it more difficult to analyze data that have been obtained in a different pipeline. It prevents the development of new types of analyzes or of new scientific questions by preventing the researchers from modifying the algorithms or inner settings of the analytic tools. It creates a 'black box' phenomenon where the results given by the analytic tools are to be trusted even though the process by which they have been obtained cannot be fully examined, tested or modified. 

The encapsulation of X!Tandem in R is a fisrt steps towards building a viable analytic pipeline for protein identification in R. It is our intention to create R objects for input and output as well as tools for data conversion (where this is possible). The goal is to make R an interesting and powerful framework for ms/ms data analysis just like it is for NGS analysis.

\section{RTandem input/output overview}

It is our intention to build R objects for the various input and output components of RTandem. But right now, the input and output of RTandem are situated 'outside of R' in XML files. The function RTandem(path) takes as input a file which sets the major  factors of the analysis. Detailed information is available on the \href{www.thegpm.org/TANDEM/index.html}{X!Tandem website}, but let's review the most important features of this file and the associated documents.

\subsection{The input file}

The main input file is used to fix the major parameters of the search, namely
\begin{enumerate}
\item The path to the output file (the file that will be used to store the results).
\item The path to the spectra file (the file that contains the data to be analysed).
\item The taxon used for the search. 
\item The path to a taxonomy file (a file that provides information on the databases to be used).
\item The path to a default-input file (a file that sets parameters that are common to many experiments).
\end{enumerate}

\subsection{The output file}

RTandem output is an XML file. The format is based on the XML language BIOML (\url{www.bioml.com}) and a full description of the format is available at \url{http://www.thegpm.org/docs/X_series_output_form.pdf}

\subsection{The spectra file}

The spectra file contains the raw data from generated by the mass spectrometer. The supported formats for the data file are 'DTA', 'PKL' and 'MGF'.

\subsection{The taxonomy file}

The taxonomy file is used to link together a taxon and one or multiple fasta files. This file is not usually modified for every experiment, rather it provides the link to all databases used and should be modified only when databases are updated or replaced.  
The supported formats for database files are 'fasta' and 'fasta.pro'.

\subsection{The default-parameter file}

The default-parameter file is used to set fine-grained parameters. For example it is used to set the mass error tolerated, or to set which types of ions will be used for calculation. This file is not usually modified for every experiment, rather, a default-parameter can be created for every instrument used (the fine-grained parameters are often instrument-specific).  
The default-parameter file also provides link to formatting files (css and xsl) that are useful for consulting the output file in a web browser. 

\section{RTandem typical usage}

Once your various input files are ready, using RTandem is very straightforward. 

Just load is as you would with any other package. 

<<Loading Rtandem,eval=FALSE>>=
library(RTandem)
@

Then launch RTandem by using the command: 
<<Launching RTandem,eval=FALSE>>=
RTandem(path_to_your_input_file)
@

A toy dataset with spectra files and fasta files as well as formatting files is included in the installation folder of RTandem. Running
<<ExampleFunction, eval=FALSE>>=
example("RTandem")
@
will create an exampleInputFile, an exampleDefaultParameter file and an exampleTaxonomy file in your working directory. Those file will contain appropriate links to the toy dataset and database in your installation directory. Then the function RTandem will be run using the newly written exampleInputFile.

\end{document}